src/lib/tecs/TECS.cpp

```diff
@@ -30,627 +30,682 @@
-				   math::trajectory::computeMaxSpeedFromDistance(
-					   _jerk_max, _vert_accel_limit, fabsf(delta_trajectory_to_target_m), 0.0f);
+	if (!PX4_ISFINITE(state.alt)) {
+		init_state.alt = 0.0f;
+	}
 
-	height_rate_target = math::constrain(height_rate_target, -target_sinkrate_m_s, target_climbrate_m_s);
+	if (!PX4_ISFINITE(state.alt_rate)) {
+		init_state.alt_rate = 0.0f;
+	}
 
-	_alt_control_traj_generator.updateDurations(height_rate_target);
-	_alt_control_traj_generator.updateTraj(_dt);
+	_alt_control_traj_generator.reset(0.0f, init_state.alt_rate, init_state.alt);
+}
 
-	_hgt_setpoint = _alt_control_traj_generator.getCurrentPosition();
-	_hgt_rate_setpoint = (_hgt_setpoint - alt_amsl) * _height_error_gain + _height_setpoint_gain_ff *
-			     _alt_control_traj_generator.getCurrentVelocity();
-	_hgt_rate_setpoint = math::constrain(_hgt_rate_setpoint, -_max_sink_rate, _max_climb_rate);
+void TECSControl::initialize()
+{
+	_ste_rate_error_filter.reset(0.0f);
+	resetIntegrals();
 }
 
-void TECS::_detect_underspeed()
+void TECSControl::update(const float dt, const Setpoint &setpoint, const Input &input, Param &param, const Flag &flag)
 {
-	if (!_detect_underspeed_enabled) {
-		_percent_undersped = 0.0f;
+	// Input checking
+	if (!TIMESTAMP_VALID(dt)) {
+		// Do not update the states and output.
+		PX4_WARN("Time intervall is not valid.");
 		return;
 	}
 
-	// this is the expected (something like standard) deviation from the airspeed setpoint that we allow the airspeed
-	// to vary in before ramping in underspeed mitigation
-	const float tas_error_bound = kTASErrorPercentage * _equivalent_airspeed_trim;
-
-	// this is the soft boundary where underspeed mitigation is ramped in
-	// NOTE: it's currently the same as the error bound, but separated here to indicate these values do not in general
-	// need to be the same
-	const float tas_underspeed_soft_bound = kTASErrorPercentage * _equivalent_airspeed_trim;
-
-	const float tas_fully_undersped = math::max(_TAS_min - tas_error_bound - tas_underspeed_soft_bound, 0.0f);
-	const float tas_starting_to_underspeed = math::max(_TAS_min - tas_error_bound, tas_fully_undersped);
+	AltitudePitchControl control_setpoint;
 
-	_percent_undersped = 1.0f - math::constrain((_tas_state - tas_fully_undersped) /
-			     math::max(tas_starting_to_underspeed - tas_fully_undersped, FLT_EPSILON), 0.0f, 1.0f);
-}
+	control_setpoint.tas_rate_setpoint = _airspeedControl(setpoint, input, param, flag);
 
-void TECS::_update_energy_estimates()
-{
-	// Calculate specific energy demands in units of (m**2/sec**2)
-	_SPE_setpoint = _hgt_setpoint * CONSTANTS_ONE_G; // potential energy
-	_SKE_setpoint = 0.5f * _TAS_setpoint_adj * _TAS_setpoint_adj; // kinetic energy
+	control_setpoint.altitude_rate_setpoint = _altitudeControl(setpoint, input, param);
 
-	// Calculate total energy error
-	_STE_error = _SPE_setpoint - _SPE_estimate + _SKE_setpoint - _SKE_estimate;
+	SpecificEnergy se{_updateEnergyBalance(control_setpoint, input)};
```

Comment:
@KonradRudin Just a suggestion, but could we rename `SpecificEngergy` to `SpecificEnergyRates` and rename `_updateEnergyBalance` to `_updateEnergyRates`.
I was slightly confused with "balance" as I associated it with the difference between potential and kinetic energy (pitch loop). What do you think?